### DISCUSSION TITLE
Using a class from other Library as CC button?

### DISCUSSION THREAD

**Discussion by josehdx** [tttapa/Control-Surface#792] [General]
Hello!

I wonder if some one can give me a hint about that to modify on my code.

I was able to set up my 2 buttons controller via BLE and now i want to go beyond.

I want to make those 2 simple buttons to have more functions, by adding "2 quick press behavior" and "press and hold" 

I found a library that has part of the code, so i am starting from small steps. I added the library and mixed with my first working BLE code but now i have lost the function of sinle press (i am staring with one button for now)

My current code looks like this:

`#include <Control_Surface.h>
#include <arduino.h>
#include <FWB.h>
BluetoothMIDI_Interface midi;

#define BP 0 //the pin where your button is connected

FWB bp;

	void OnClick(int pin)
	{
	CCButton bp[] {
  						{0, {MIDI_CC::General_Purpose_Controller_1, CHANNEL_1} },
	};
	
	}

void setup()
 {

	Control_Surface.begin(); // Initialize Control Surface
 	bp.Configure(BP);
 	bp.OnClick = OnClick;
	
 }`

My tests so far was  to try with the orifinal class "CCButton button" and also changed the button class for bp and with BP but it does not compile.

Hope you can help!

**Comment by josehdx** [comment]
looks that i did not copied the loop section


`void loop() {
  Control_Surface.loop(); // Update the Control Surface

  //	// Test button state
	bp.CheckBP();
}`

**Comment by tttapa** [comment]
You don't need an external library, e.g.:

```cpp
#include <Control_Surface.h>
#include <AH/Hardware/MultiPurposeButton.hpp>

BluetoothMIDI_Interface midi;
 
// Create a Button object that reads a push button connected to pin 0:
MultiPurposeButton btn{0};

void setup() {
  Control_Surface.begin();
  btn.setLongPressDelay(1000);
  btn.setMultiPressDelay(400);
  btn.begin();
}

void loop() {
  switch (btn.update()) {
    case btn.None: break;
    case btn.PressStart: break;
    case btn.ShortPressRelease:
      Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_1}, 0);
      break;
    case btn.LongPress: /* ... */ break;
    case btn.LongPressRelease: /* ... */ break;
    case btn.MultiPress: break;
    case btn.MultiPressDone: break;
  }
  Control_Surface.loop();
}
```

**Reply by josehdx** [reply]
OMG, this library has everything :D i will check it now

**Reply by tttapa** [reply]
See also: https://tttapa.github.io/Control-Surface-doc/Doxygen/db/d45/3_8MultiPurposeButton_8ino-example.html

**Reply by josehdx** [reply]
I have checked that last library before online, and actually, i was trying  a new example with  the code you pasted, but the reultts might be a little different for what i was expecting. What i am looking for, is to "latch" the new midi command i try to use after a double or hold press, i mean, if i set the starting MIDI message as for example MIDIcc16, i want to change it to MIDI 17 uf i keep press the button 0, but the code looks like is a momentary "filter", i think i need to use some condtionals to change the behavior. Additionally i want to add a second button, and when i added on this way, pin 35 did not work: 



`case btn.ShortPressRelease:
      Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_1}, 0);
      Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_2}, 35);
         break;`

Thanks

**Reply by tttapa** [reply]
> What i am looking for, is to "latch" the new midi command i try to use after a double or hold press, i mean, if i set the starting MIDI message as for example MIDIcc16, i want to change it to MIDI 17 uf i keep press the button 0

Then you'll have to keep the address in a variable, and change that variable on a long press (in the `/* ... */`). For instructions on how to store MIDI addresses in variables, see https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/df7/midi-tutorial.html#midi_md-midi-addresses.

> but the code looks like is a momentary "filter", i think i need to use some condtionals to change the behavior.

Indeed, you'll have to add some conditionals.

> Additionally i want to add a second button, and when i added on this way, pin 35 did not work:
>
> `case btn.ShortPressRelease: Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_1}, 0); Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_2}, 35); break;`

The zero you replaced by 35 is not the pin number, it's the CC value, see the documentation for `sendControlChange`: https://tttapa.github.io/Control-Surface-doc/Doxygen/dc/d45/classControl__Surface__.html#aa0e05cc39766771989374c1dd26b77a4  

The pin number is specified as a constructor argument of `MultiPurposeButton`, on line 7.

**Comment by josehdx** [comment]
Checking both classes, is obvious  that they have different properties and on the case of MultiPurposeButton, different events. And if the goal is to use MIDI addresses instead of the MultiPurposeButton, should i create a new class based on the MIDI address, with some of the propoerties, or is it possible somehow to combine both in order to make the "case" work? 

And also, how should i use [AH/Timing/MillisMicrosTimer.hpp] to set a delay to the "first single press" in order to avoid the first MIDI (0 to 127) printing and allow to do the second press with out interruptions?

this value "btn.setMultiPressDelay(400); " is for the time between pressing as far as i can tell... and the delay should be higher that (400)


Thanks
